benchmarks/encryption.py

```diff
@@ -0,0 +1,330 @@
+from abc import ABC
+from typing import IO
+from typing import Iterable
+from os import urandom
+from contextlib import contextmanager
+from tempfile import NamedTemporaryFile
+from base64 import b64encode
+
+
+from cryptography.hazmat.primitives.ciphers import Cipher
+from cryptography.hazmat.primitives.ciphers.algorithms import AES
+from cryptography.hazmat.primitives.ciphers.modes import CTR
+from cryptography.hazmat.primitives.hmac import HMAC
+from cryptography.hazmat.primitives.hashes import SHA256
+from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
+from cryptography.hazmat.backends import default_backend
+
+from itertools import tee
+from os import urandom
+
+PASSWORD_DERIVE_ITER = 100000
+
+
+class KeyDerive(object):
+
+    def __init__(self, password, iter, hash, derive, backend, salt):
+
+        self._password = password
+        self._iterations = iter
+        self._hash = hash
+        self._derive = derive
+        self._backend = backend
+        self._salt = salt
+
+        self._length = 16
+
+    def _generate_key(self):
+
+        kdf = self._derive(
+            algorithm=self._hash(),
+            length=self._length,
+            salt=self._salt,
+            iterations=self._iterations,
+            backend=self._backend()
+        )
+
+        return kdf.derive(self._password)
+
+    def _verify(self, k):
+
+        kdf = self._derive(
+            algorithm=self._hash(),
+            length=self._length,
+            salt=self._salt,
+            iterations=self._iterations,
+            backend=self._backend()
+        )
+
+        kdf.verify(self._password, k)
+
+    def __call__(self):
+
+        key = self._generate_key()
+        self._verify(key)
+        return key
+
+
+class Encrypt(object):
+
+    def __init__(self, key, hmac_key, hmac, hash, cipher, algorithm, mode, backend, salt, hmac_salt):
+
+        self._key = key
+        self._hmac_key = hmac_key
+        self._hmac = hmac
+        self._hash = hash
+        self._cipher = cipher
+        self._algorithm = algorithm
+        self._mode = mode
+        self._backend = backend
+        self._salt = salt
+        self._hmac_salt = hmac_salt
+
+        self._iv = urandom(16)
+
+    def __call__(self, data):
+
+        h = self._hmac(
+            self._hmac_key,
+            self._hash(),
+            backend=self._backend()
+        )
+
+        ciph = self._cipher(
+            self._algorithm(self._key),
+            self._mode(self._iv),
+            backend=self._backend()
+        )
+
+        encryptor = ciph.encryptor()
+        iv_sent = False
+
+        for raw in data:
+            out = encryptor.update(raw)
+
+            if not iv_sent:
+                out = self._iv + self._salt + self._hmac_salt + out
+                iv_sent = True
+
+            h.update(out)
```

Comment:
Could you expand on this variable name to make the code more explicit?

Reply:
For sure, will do!